### PR TITLE
Fix inset locator renderer fallback (swev-id: matplotlib__matplotlib-26291)

### LINKED ISSUE #67: [Bug] inset_axes fails under inline backend with bbox_inches=tight (renderer None)
### User Request

[Bug]: Error while creating inset axes using `mpl_toolkits.axes_grid1.inset_locator.inset_axes`

Bug summary:
Unable to create the inset axes in a plot using the code (following the first example on the website as posted [here](https://matplotlib.org/stable/gallery/axes_grid1/inset_locator_demo.html)).

Code for reproduction:
```python
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1.inset_locator import inset_axes

fig, (ax, ax2) = plt.subplots(1, 2, figsize=[5.5, 2.8])
axins = inset_axes(ax, width=1.3, height=0.9)
plt.show()
```

Actual outcome (stack trace excerpt):
```Python
AttributeError: 'NoneType' object has no attribute '_get_renderer'
... at matplotlib/offsetbox.py:399 in OffsetBox.get_window_extent
... called from mpl_toolkits/axes_grid1/inset_locator.py:73 in AnchoredLocatorBase.__call__
... triggered via matplotlib/_tight_bbox.py during inline backend bbox_inches='tight'
```

Expected outcome:
An empty box towards the top right of the first subplot (with axes `ax`), as shown in the demo.

Environment:
- OS: Arch linux: 6.4.2-arch1-1
- Matplotlib Version: 3.7.2
- Backend: module://matplotlib_inline.backend_inline
- Python: 3.8.17
- Jupyter lab: 3.6.5
- Installation: conda

Notes:
- Inline backend applies `bbox="tight"` to figure display.
- The axes created by `axes_grid1.insetlocator.inset_axes` are not compatible with tight layout; however, `get_window_extent` should still work.

Task reference: ID 284

---

### Researcher Specification (Emerson Gray)

Overview:
Creating inset axes via `mpl_toolkits.axes_grid1.inset_locator.inset_axes` fails under IPython inline backend when `bbox="tight"` is applied implicitly. The failure occurs because `AnchoredLocatorBase.__call__` forwards `renderer=None` to `OffsetBox.get_window_extent`, which then tries to use `self.figure._get_renderer()`. Since the locator object is not attached to a figure (`self.figure` is `None`), this raises `AttributeError`.

Minimal fix: Ensure `AnchoredLocatorBase.__call__` obtains a valid renderer from the parent Axes’ figure when the passed `renderer` is `None`.

Files and exact change:
- Target file: `lib/mpl_toolkits/axes_grid1/inset_locator.py`

Patch diff chunk:
```diff
*** Begin Patch
*** Update File: lib/mpl_toolkits/axes_grid1/inset_locator.py
@@
 class AnchoredLocatorBase(AnchoredOffsetbox):
@@
     def draw(self, renderer):
         raise RuntimeError("No draw method should be called")
 
     def __call__(self, ax, renderer):
+        # Some callers (e.g., tight bbox adjustment or inset axes setup)
+        # invoke locators with renderer=None. OffsetBox.get_window_extent
+        # will fallback to self.figure._get_renderer() if renderer is None,
+        # but this locator is not attached to a Figure (self.figure is None).
+        # Ensure a valid renderer is used.
+        if renderer is None:
+            renderer = ax.figure._get_renderer()
         self.axes = ax
         bbox = self.get_window_extent(renderer)
         px, py = self.get_offset(bbox.width, bbox.height, 0, 0, renderer)
         bbox_canvas = Bbox.from_bounds(px, py, bbox.width, bbox.height)
         tr = ax.figure.transSubfigure.inverted()
         return TransformedBbox(bbox_canvas, tr)
*** End Patch
```

Compatibility and behavior:
- Inline backend with implicit `bbox_inches='tight'`: fixed.
- Non-inline backends and non-tight usage: unaffected.
- Other `axes_grid1` locators benefit similarly.
- We do not attach the locator to the figure; passing a valid renderer is sufficient.

Testing plan:
- Manual repro (pre-fix failure, post-fix success):
```python
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1.inset_locator import inset_axes
fig, (ax, ax2) = plt.subplots(1, 2, figsize=(5.5, 2.8))
axins = inset_axes(ax, width=1.3, height=0.9)
fig.savefig('out.png', bbox_inches='tight')
```
- Unit test proposal: Add a smoketest to `lib/mpl_toolkits/axes_grid1/tests/test_axes_grid1.py` that creates an inset axes and saves with `bbox_inches='tight'` to ensure no exception.

Edge cases:
- `tight_layout`: existing warnings unchanged; fix prevents error.
- `constrained_layout`: unaffected; the renderer derivation is safe.

Reason:
Prevent `OffsetBox.get_window_extent` from using `self.figure` when the locator is not attached to a figure, avoiding `AttributeError` during inline backend tight bbox processing.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- ensure `AnchoredLocatorBase.__call__` acquires a renderer if one was not provided
- prevent AttributeError when inline backend implicitly requests `bbox_inches='tight'`
- keep existing locator behavior otherwise unchanged

## Observed Failure
- `AttributeError: 'NoneType' object has no attribute '_get_renderer'` when `inset_axes` is used under the inline backend with tight bounding box rendering

## Reproduction
1. Run the inline smoketest snippet from Issue #67 to see the AttributeError
2. With this branch, rerun the same snippet and verify it completes and saves the figure
3. Confirm `Axes.inset_axes` still works with `bbox_inches='tight'`

Fixes #67

### CODE REVIEW

**Comment by casey-brooks** [comment]
### Test & Lint Summary
LD_LIBRARY_PATH=/root/.nix-profile/lib:/nix/store/qipd93x9gjyiygqk673rd2ssnf8y7jj0-gcc-14.3.0-lib/lib:/nix/store/f8w1i7yisixb9hivzbk0l4ixmf67fjqr-gcc-14.3.0-libgcc/lib:$LD_LIBRARY_PATH MPLBACKEND=Agg .venv/bin/python - <<'PY'
import matplotlib
import matplotlib.pyplot as plt
from mpl_toolkits.axes_grid1.inset_locator import inset_axes
fig, (ax, ax2) = plt.subplots(1, 2, figsize=(5.5, 2.8))
axins = inset_axes(ax, width=1.3, height=0.9)
fig.savefig('out.png', bbox_inches='tight')
print('smoketest1 ok')
PY
→ passed

LD_LIBRARY_PATH=/root/.nix-profile/lib:/nix/store/qipd93x9gjyiygqk673rd2ssnf8y7jj0-gcc-14.3.0-lib/lib:/nix/store/f8w1i7yisixb9hivzbk0l4ixmf67fjqr-gcc-14.3.0-libgcc/lib:$LD_LIBRARY_PATH MPLBACKEND=Agg .venv/bin/python - <<'PY'
import matplotlib
import matplotlib.pyplot as plt
fig, ax = plt.subplots()
axins = ax.inset_axes([0.6, 0.6, 0.3, 0.3])
fig.savefig('axes_inset.png', bbox_inches='tight')
print('smoketest2 ok')
PY
→ passed

.venv/bin/flake8 lib/mpl_toolkits/axes_grid1/inset_locator.py
→ no issues

File: lib/mpl_toolkits/axes_grid1/inset_locator.py [hunk 1]
```diff
@@ -69,6 +69,8 @@ def draw(self, renderer):
         raise RuntimeError("No draw method should be called")
 
     def __call__(self, ax, renderer):
+        if renderer is None:
+            renderer = ax.figure._get_renderer()
```
Comment (noa-lucent) [inline lib/mpl_toolkits/axes_grid1/inset_locator.py:73]: [major] Thanks for untangling the renderer hand-off here. Can we add a regression test that exercises the renderer=None path (e.g. reproduce the inline backend + bbox_inches='tight' scenario from #67)? Without a test, this could easily regress again. A small smoketest in lib/mpl_toolkits/axes_grid1/tests that saves a figure with inset_axes and bbox_inches='tight' should cover it.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for fixing the renderer hand-off. We still need a regression test to make sure the inline backend + bbox_inches='tight' path is covered.

**Comment by casey-brooks** [comment]
Added  covering the inline tight-bbox path. Verified locally with:



The test passed.

**Review by noa-lucent — APPROVED** [review]
Regression smoketest looks good; thanks for adding it.
